src/coreclr/gc/gc.cpp

```diff
@@ -8368,29 +8391,262 @@ void destroy_card_table (uint32_t* c_table)
     dprintf (2, ("Table Virtual Free : %Ix", (size_t)&card_table_refcount(c_table)));
 }
 
-uint32_t* gc_heap::make_card_table (uint8_t* start, uint8_t* end)
+void gc_heap::get_card_table_element_sizes (uint8_t* start, uint8_t* end, size_t sizes[total_bookkeeping_elements])
 {
-    assert (g_gc_lowest_address == start);
-    assert (g_gc_highest_address == end);
+    sizes[card_table_element] = size_card_of (start, end);
+    sizes[brick_table_element] = size_brick_of (start, end);
 
-    uint32_t virtual_reserve_flags = VirtualReserveFlags::None;
+    sizes[brick_table_element + 1] = 0;
+#ifdef CARD_BUNDLE
+    if (can_use_write_watch_for_card_table())
+    {
+        sizes[card_bundle_table_element] = size_card_bundle_of (start, end);
+    }
+#endif //CARD_BUNDLE
 
-    size_t bs = size_brick_of (start, end);
-    size_t cs = size_card_of (start, end);
+    sizes[seg_mapping_table_element - 1] = 0;
+#ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+    if (gc_can_use_concurrent)
+    {
+        sizes[software_write_watch_table_element] = SoftwareWriteWatch::GetTableByteSize(start, end);
+    }
+#endif // FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+
+    sizes[seg_mapping_table_element] = size_seg_mapping_table_of (start, end);
+
+    sizes[mark_array_element] = 0;
 #ifdef BACKGROUND_GC
-    size_t ms = (gc_can_use_concurrent ?
-                 size_mark_array_of (start, end) :
-                 0);
-#else
-    size_t ms = 0;
+    if (gc_can_use_concurrent)
+    {
+        sizes[mark_array_element] = size_mark_array_of (start, end);
+    } 
 #endif //BACKGROUND_GC
+}
+
+void gc_heap::get_card_table_element_layout (uint8_t* start, uint8_t* end, size_t layout[total_bookkeeping_elements + 1])
+{
+    size_t last;
+    size_t sizes[total_bookkeeping_elements];
+    get_card_table_element_sizes(start, end, sizes);
+
+    layout[card_table_element] = last = sizeof(card_table_info);
+    layout[brick_table_element] = last = last + sizes[card_table_element];
+    int missed_element = brick_table_element + 1;
+#ifdef CARD_BUNDLE
+    layout[card_bundle_table_element] = last = last + sizes[brick_table_element];
```

Comment:
could be rewritten to 
`layout[card_bundle_table_element] = layout[card_bundle_table_element - 1] + sizes[brick_table_element];`
so you don't need to maintain a last?